File: packs/dsa-trees-py/challenges/03-min-depth.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import min_depth


def test_balanced_tree():
    root = build_tree([3, 9, 20, None, None, 15, 7])
    assert min_depth(root) == 2


def test_right_skewed():
    root = build_tree([2, None, 3, None, 4, None, 5, None, 6])
    assert min_depth(root) == 5


def test_empty_tree():
    assert min_depth(None) == 0


def test_single_node():
    root = build_tree([1])
    assert min_depth(root) == 1


def test_left_leaf_shorter():
    root = build_tree([1, 2, 3, None, None, 4, 5])
    assert min_depth(root) == 2

File: packs/dsa-trees-ts/challenges/03-min-depth.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '03-min-depth.json');

describe('Minimum Depth', () => {
  let minDepth: (root: TreeNode | null) => number;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ minDepth: typeof minDepth }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    minDepth = mod.minDepth;
  });

  afterAll(() => cleanup());

  it('returns 2 for a balanced tree', () => {
    expect(minDepth(buildTree([3, 9, 20, null, null, 15, 7]))).toBe(2);
  });

  it('returns 5 for right-skewed tree', () => {
    expect(minDepth(buildTree([2, null, 3, null, 4, null, 5, null, 6]))).toBe(5);
  });

  it('returns 0 for empty tree', () => {
    expect(minDepth(null)).toBe(0);
  });

  it('returns 1 for single node', () => {
    expect(minDepth(buildTree([1]))).toBe(1);
  });

  it('returns 2 when left leaf is shorter', () => {
    expect(minDepth(buildTree([1, 2, 3, null, null, 4, 5]))).toBe(2);
  });
});
